pandas/core/strings.py

```diff
@@ -2307,23 +2324,35 @@ def cat(self, others=None, sep=None, na_rep=None, join=None):
         na_masks = np.array([isna(x) for x in all_cols])
         union_mask = np.logical_or.reduce(na_masks, axis=0)
 
-        if na_rep is None and union_mask.any():
-            # no na_rep means NaNs for all rows where any column has a NaN
-            # only necessary if there are actually any NaNs
-            result = np.empty(len(data), dtype=object)
-            np.putmask(result, union_mask, np.nan)
-
-            not_masked = ~union_mask
-            result[not_masked] = cat_core([x[not_masked] for x in all_cols],
-                                          sep)
-        elif na_rep is not None and union_mask.any():
-            # fill NaNs with na_rep in case there are actually any NaNs
-            all_cols = [np.where(nm, na_rep, col)
-                        for nm, col in zip(na_masks, all_cols)]
-            result = cat_core(all_cols, sep)
-        else:
-            # no NaNs - can just concatenate
-            result = cat_core(all_cols, sep)
+        # if there are any non-string, non-null values hidden within an object
+        # dtype, cat_core will fail; catch error and return with better message
+        try:
+            if na_rep is None and union_mask.any():
+                # no na_rep means NaNs for all rows where any column has a NaN
+                # only necessary if there are actually any NaNs
+                result = np.empty(len(data), dtype=object)
+                np.putmask(result, union_mask, np.nan)
+
+                not_masked = ~union_mask
+                result[not_masked] = cat_core([x[not_masked]
+                                               for x in all_cols], sep)
+            elif na_rep is not None and union_mask.any():
+                # fill NaNs with na_rep in case there are actually any NaNs
+                all_cols = [np.where(nm, na_rep, col)
+                            for nm, col in zip(na_masks, all_cols)]
+                result = cat_core(all_cols, sep)
+            else:
+                # no NaNs - can just concatenate
+                result = cat_core(all_cols, sep)
+        except TypeError as exc:
+            if re.match((r'can only concatenate str \(not [\w\"]+) to str'
```

Comment:
if we reach here, we already have a user-facing TypeError? and we are just replacing the message.

are any other TypeErrors expected, i'd be inclined to not check the message check.

if it is necessary, maybe have a `_is_not_string_exception` in pandas\core\dtypes\common.py similar to `_is_unorderable_exception` rather than have the condition here.

Reply:
> are any other TypeErrors expected, i'd be inclined to not check the message check.

Fair enough, that's a not gonna happen.